File: clients/javascript/src/transport/http.ts
import { ConnectionError } from "../errors";
import type { Transport, TransportExecuteOptions, TransportResponse } from "./base";
import type { Logger } from "../utils/logger";

export interface HttpTransportOptions {
  endpoint: string;
  readTimeoutMs?: number;
  fetchImpl?: typeof fetch;
  logger: Logger;
}

export class HttpTransport implements Transport {
  readonly kind = "http" as const;

  private readonly endpoint: string;
  private readonly readTimeoutMs: number;
  private readonly fetchImpl: typeof fetch;
  private readonly logger: Logger;

  constructor(options: HttpTransportOptions) {
    this.endpoint = options.endpoint.replace(/\/$/, "") + "/command";
    this.readTimeoutMs = options.readTimeoutMs ?? 60_000;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch;
    this.logger = options.logger;

    if (!this.fetchImpl) {
      throw new ConnectionError("Global fetch is not available. Provide a fetch implementation in the client options.");
    }
  }

  async execute(command: string, options?: TransportExecuteOptions): Promise<TransportResponse> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.readTimeoutMs);
    const headers = {
      "Content-Type": "text/plain",
      ...(options?.headers ?? {})
    };

    try {
      this.logger.debug(
        `HTTP execute -> ${this.endpoint} (bytes=${command.length})`
      );
      const response = await this.fetchImpl(this.endpoint, {
        method: "POST",
        body: command,
        headers,
        signal: controller.signal
      });

      const body = await response.text();
      this.logger.debug(
        `HTTP response <- ${this.endpoint} status=${response.status} bytes=${body.length}`
      );
      return {
        status: response.status,
        body,
        headers: this.normalizeHeaders(response.headers)
      };
    } catch (error) {
      if ((error as Error).name === "AbortError") {
        this.logger.warn(
          `HTTP request to ${this.endpoint} aborted after ${this.readTimeoutMs}ms`
        );
        throw new ConnectionError(`HTTP request timeout after ${this.readTimeoutMs}ms`);
      }
      const message = error instanceof Error ? error.message : String(error);
      this.logger.error(`HTTP request failed: ${message}`);
      throw new ConnectionError(`Cannot connect to server at ${this.endpoint}: ${message}`);
    } finally {
      clearTimeout(timeoutId);
    }
  }

  // Stateless transport, nothing to close
  async close(): Promise<void> {
    return Promise.resolve();
  }

  private normalizeHeaders(headers: Headers): Record<string, string> {
    const normalized: Record<string, string> = {};
    headers.forEach((value, key) => {
      normalized[key.toLowerCase()] = value;
    });
    return normalized;
  }
}

File: clients/python/src/sneldb_client/transport/http.py
"""HTTP transport built on top of httpx."""

from __future__ import annotations

import httpx

from ..errors import ConnectionError
from ..logger import BoundLogger, create_logger
from .base import Transport, TransportResponse


class HttpTransport:
    kind: Transport.Kind = "http"

    def __init__(
        self,
        endpoint: str,
        *,
        read_timeout: float = 60.0,
        client: httpx.Client | None = None,
        logger: BoundLogger | None = None,
    ) -> None:
        self._endpoint = endpoint.rstrip("/") + "/command"
        self._read_timeout = read_timeout
        self._client = client or httpx.Client(timeout=httpx.Timeout(read_timeout))
        self._owns_client = client is None
        self._logger = (logger or create_logger()).child("http")

    def execute(self, command: str, headers: dict[str, str] | None = None) -> TransportResponse:
        payload = command
        request_headers = {"Content-Type": "text/plain"}
        if headers:
            request_headers.update(headers)

        try:
            self._logger.debug("HTTP POST %s bytes=%d", self._endpoint, len(payload))
            response = self._client.post(self._endpoint, content=payload, headers=request_headers)
            body = response.content
            self._logger.debug(
                "HTTP <- %s status=%s bytes=%d",
                self._endpoint,
                response.status_code,
                len(body),
            )
            return TransportResponse(
                status=response.status_code,
                body=body,
                headers={k.lower(): v for k, v in response.headers.items()},
            )
        except httpx.TimeoutException as exc:
            raise ConnectionError(f"HTTP request timeout after {self._read_timeout}s") from exc
        except httpx.RequestError as exc:
            raise ConnectionError(f"Cannot connect to {self._endpoint}: {exc}") from exc

    def close(self) -> None:
        if self._owns_client:
            self._client.close()


__all__ = ["HttpTransport"]
